fuel-benches/src/lib.rs

```diff
@@ -0,0 +1,268 @@
+}
+
+impl VmBench {
+    pub fn new(instruction: Opcode) -> Self {
+        Self {
+            params: ConsensusParameters::default(),
+            gas_price: 0,
+            gas_limit: 1_000_000,
+            maturity: 0,
+            height: 0,
+            prepare_script: vec![],
+            data: vec![],
+            inputs: vec![],
+            outputs: vec![],
+            witnesses: vec![],
+            db: None,
+            instruction,
+        }
+    }
+
+    pub fn contract<R>(rng: &mut R, instruction: Opcode) -> io::Result<Self>
+    where
+        R: Rng,
+    {
+        let bench = Self::new(Opcode::CALL(0x10, REG_ZERO, 0x11, 0x12));
+
+        let program = iter::once(instruction)
+            .chain(iter::once(Opcode::RET(REG_ONE)))
+            .collect::<Vec<u8>>();
+
+        let program = Witness::from(program);
+
+        let salt = rng.gen();
+
+        let contract = Contract::from(program.as_ref());
+        let contract_root = contract.root();
+        let state_root = Contract::default_state_root();
+        let contract = contract.id(&salt, &contract_root, &state_root);
+
+        let utxo_id = rng.gen();
+        let balance_root = rng.gen();
+        let tx_pointer = rng.gen();
+
+        let input =
+            Input::contract(utxo_id, balance_root, state_root, tx_pointer, contract);
+        let output = Output::contract_created(contract, state_root);
+
+        let bytecode_witness = 0;
+        let tx = Transaction::create(
+            bench.gas_price,
+            bench.gas_limit,
+            bench.maturity,
+            bytecode_witness,
+            salt,
+            vec![],
+            vec![],
+            vec![output],
+            vec![program],
+        )
+        .check(bench.height, &bench.params)?;
+
+        let output = Output::contract(0, rng.gen(), rng.gen());
+
+        let mut db = new_db()?;
+
+        let mut txtor = Transactor::new(&mut db, bench.params);
+
+        txtor.transact(tx);
+
+        let data = contract
+            .iter()
+            .copied()
+            .chain((0 as Word).to_be_bytes().iter().copied())
+            .chain((0 as Word).to_be_bytes().iter().copied())
+            .chain(AssetId::default().iter().copied())
+            .collect();
+
+        let prepare_script = vec![
+            Opcode::gtf(0x10, 0x00, GTFArgs::ScriptData),
+            Opcode::ADDI(0x11, 0x10, ContractId::LEN as Immediate12),
+            Opcode::ADDI(0x11, 0x11, WORD_SIZE as Immediate12),
+            Opcode::ADDI(0x11, 0x11, WORD_SIZE as Immediate12),
+            Opcode::MOVI(0x12, 100_000),
+        ];
+
+        Ok(bench
+            .with_db(db)
+            .with_data(data)
+            .with_input(input)
+            .with_output(output)
+            .with_prepare_script(prepare_script))
+    }
+
+    pub fn with_db(mut self, db: Database) -> Self {
+        self.db.replace(db);
+        self
+    }
+
+    pub fn with_params(mut self, params: ConsensusParameters) -> Self {
```

Comment:
nit: I dislike having setters like this for public fields, especially when there is no need to use builder pattern.

Reply:
The idea of these are to avoid having to declare every and each attribute for tests that requires only one attribute (ie instruction).

wdyt?

Reply:
I feel like I'm missing something here, what's wrong with this?

```rust
let mut vm_bench = VmBench::new(Opcode::NOP);
vm_bench.params = params;
self.gas_price = gas_price;
```

Reply:
Seems like it's mainly to enable defining benches inline with the criterion run fn's, such as:

```rust
    run_group(
        &mut group,
        "mlog",
        VmBench::new(Opcode::MLOG(0x10, 0x11, 0x12)).with_prepare_script(vec![
            Opcode::MOVI(0x11, 100000),
            Opcode::MOVI(0x12, 27),
        ]),
    );
```

Although the same thing could be accomplished with:

```rust
    run_group(
        &mut group,
        "mlog",
        VmBench {
           prepare_script: vec![
             Opcode::MOVI(0x11, 100000),
             Opcode::MOVI(0x12, 27),
           ])
           ..Self::new(Opcode::MLOG(0x10, 0x11, 0x12))
        }
    );
```